File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    # 친구 이름을 숫자로 변경
    dict = {}
    for i in range(len(friends)):
        dict[friends[i]] = [i, 0]

    # 선물을 준 기록을 2차원 배열로 정리
    table = [[0] * len(friends) for _ in range(len(friends))]
    for gift in gifts:
        a, b = gift.split()
        table[dict[a][0]][dict[b][0]] += 1

        # 선물지수 갱신
        dict[a][1] += 1
        dict[b][1] -= 1
    
    # 각 friend마다 다음달에 받을 선물 개수(cnt)를 계산
    answer = 0
    for a in friends:
        cnt = 0
        for b in friends:
            if a == b : continue
            elif table[dict[a][0]][dict[b][0]] > table[dict[b][0]][dict[a][0]]:
                cnt += 1
            elif table[dict[a][0]][dict[b][0]] == table[dict[b][0]][dict[a][0]]:
                if dict[a][1] > dict[b][1]:
                    cnt += 1
        answer = cnt if cnt > answer else answer
    
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
// 배열 선언법, HashMap 선언법 체크
// dict.put(key, value) 형태로 값 주입
// dict.get(key) 형태로 값 획득

import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int n = friends.length;
        HashMap<String, Integer> dict = new HashMap<>();
        int[] degree = new int[n];
        int[][] graph = new int[n][n];
        
        // 해시맵에 값 주입
        for (int i = 0; i < n; i++) {
            dict.put(friends[i], i);
        }
        
        // 선물 기록으로 지금까지 주고받은 기록과 지수 계산
        for (String gift : gifts) {
            String[] giftName = gift.split(" ");
            degree[dict.get(giftName[0])]++;
            degree[dict.get(giftName[1])]--;
            graph[dict.get(giftName[0])][dict.get(giftName[1])]++;    
        }
        
        for (int i =0; i < n; i++) {
            int num = 0;
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                if (graph[i][j] > graph[j][i]) num++;   // 내가 더 준 경우
                if (graph[i][j] == graph[j][i]          // 내가 선물지수가 더 큰 경우
                   && degree[i] > degree[j]) num++;
            }
            
            if (answer < num) answer = num;
        }
        return answer;
    }
}